src/restLink.ts

```diff
@@ -703,14 +748,40 @@ const resolver: Resolver = async (
           linkLevelNameDenormalizer ||
           noOpNameNormalizer,
       );
+
+      const bothSerializersProvided =
+        bodySerializerKey != null && bodySerializer != null;
+
+      if (bothSerializersProvided) {
+        throw new Error(
+          'One and only one of ("bodySerializer" | "bodySerializerKey") must be set in the @rest() directive. ' +
+            'This request had both',
+        );
+      }
+
+      if (bodySerializerKey && !serializers.hasOwnProperty(bodySerializerKey)) {
+        throw new Error(
+          '"bodySerializerKey" must correspond to configured serializer. ' +
+            `Please make sure to specify a serializer called ${bodySerializerKey} in the "additionalSerializers" property of the RestLink.`,
+        );
+      }
+
+      const serializedBody = (bodySerializer
+        ? bodySerializer
+        : serializers[bodySerializerKey || DEFAULT_SERIALIZER_KEY])(body);
+
+      body = serializedBody.body;
+      additionalHeaders = new Headers(serializedBody.headers);
     }
 
     validateRequestMethodForOperationType(method, operationType || 'query');
     return await (customFetch || fetch)(`${uri}${pathWithParams}`, {
       credentials,
       method,
-      headers,
-      body: body && JSON.stringify(body),
+      headers: additionalHeaders
+        ? headersMergePolicy(additionalHeaders, headers)
```

Comment:
This I'm actually not to sure about @fbartho, if I switch the order here, which I initially tried to do, it broke some other non-related tests. Doing this keeps all tests passing, but I guess we should decide *what* the expected behavior of this should be? I assume the serializer should in general get the final say here, but I think this might be doing the opposite no?